src/ZkFold/Base/Protocol/ARK/Plonk.hs

```diff
@@ -41,14 +43,28 @@ type G2 = Point BLS12_381_G2
     Additionally, we don't want this library to depend on Cardano libraries.
 -}
 
+-- TODO: disallow `d` that are not powers of 2
 data Plonk (d :: Natural) (n :: Natural) t = Plonk F F F (Vector n Natural) (ArithmeticCircuit 1 F) F
     deriving (Show)
+
 -- TODO (Issue #25): make a proper implementation of Arbitrary
-instance Arbitrary (Plonk d n t) where
+instance (KnownNat d, KnownNat n) => Arbitrary (Plonk d n t) where
     arbitrary = do
-        let (omega, k1, k2) = getParams 5
         ac <- arbitrary
-        Plonk omega k1 k2 (Vector [1, 2]) ac <$> arbitrary
+        let fullInp = length . inputVariables $ ac
+        vecPubInp <- genSubset (return []) (value @n) fullInp
+        let (omega, k1, k2) = getParams $ ceiling @Double @Natural $ logBase 2 $ fromIntegral (value @d)
+        Plonk omega k1 k2 (Vector vecPubInp) ac <$> arbitrary
+
+genSubset :: Gen [Natural] -> Natural -> Natural -> Gen [Natural]
```

Comment:
type of the first argument of this function can be safely changed to `[Natural]`

Reply:
Or even better to `Set Natural`, as can be inferred from its usage